packages/ember-application/lib/system/application.js

```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
   },
 
   /**
-    Calling initialize manually is not supported.
-
-    Please see Ember.Application#advanceReadiness and
-    Ember.Application#deferReadiness.
+    Initialize the application and return a promise that resolves with the `Ember.Application`
+    object when the boot process is complete.
 
-    @private
-    @deprecated
-    @method initialize
-   **/
-  initialize() {
-    deprecate('Calling initialize manually is not supported. Please see Ember.Application#advanceReadiness and Ember.Application#deferReadiness');
-  },
+    Run any application initializers and run the application load hook. These hooks may
+    choose to defer readiness. For example, an authentication hook might want to defer
+    readiness until the auth token has been retrieved.
 
-  /**
-    Initialize the application. This happens automatically.
+    By default, this method is called automatically on "DOM ready"; however, if autoboot
+    is disabled, you would need to call this method yourself. Calling this method before
+    "DOM ready" is not recommended.
 
-    Run any initializers and run the application load hook. These hooks may
-    choose to defer readiness. For example, an authentication hook might want
-    to defer readiness until the auth token has been retrieved.
-
-    @private
-    @method domReady
+    @public
+    @method boot
+    @return {Promise<Ember.Application,Error>}
   */
-  domReady() {
-    if (this.isDestroyed) { return; }
+  boot() {
+    if (this._bootPromise) { return this._bootPromise; }
 
-    this.boot();
+    try {
+      this._bootSync();
+    } catch(_) {
+      // Ignore th error: in the asynchronous boot path, the error is already reflected
+      // in the promise rejection
+    }
 
-    return this;
+    return this._bootPromise;
   },
 
-  boot() {
-    if (this._bootPromise) { return this._bootPromise; }
+  /**
+    @private
+  */
+  _booted: false,
 
-    var defer = new Ember.RSVP.defer();
-    this._bootPromise = defer.promise;
-    this._bootResolver = defer;
+  /**
+    Unfortunately, a lot of existing code assumes the booting process is
+    "synchronous". Specifically, a lot of tests assumes the last call to
+    `app.advanceReadiness()` or `app.reset()` will result in the app being
+    fully-booted when the current runloop completes.
 
-    this.runInitializers();
-    runLoadHooks('application', this);
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an app "synchronously".
 
-    this.advanceReadiness();
+    @private
+  */
+  _bootSync() {
```

Comment:
It is a stop-gap, but we probably won't be able to completely remove it until 3.0; we should start fixing our tests and somehow deprecate them though.

Basically, we have tests that looks like this:

```javascript
test('...', function() {
  var App = ...;

  run(App, 'advanceReadiness');

  nowSynchronouslyDoStuffSinceAppIsFullyBooted();
});
```

or 

```javascript
test('...', function() {
  var App = ...;

  // ...

  run(App, 'reset');

  nowSynchronouslyDoStuffSinceAppIsFullyBooted();
});
```

(I should probably edit the comment to not say "most of them are already deprecated", turns out we still support globals mode fully?)